src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,376 @@
+# considered as word chars, thus not matched by "\w" category in regex
+HAS_WEIRD_WORD_CHARS = (2, 7) <= sys.version_info[:2] < (3, 4)
+UNICODE_WEIRD_NONWORD_CHARS = set(u'\U00012432\U00012433\U00012456\U00012457')
+
+
+class Context(object):
+    __slots__ = ['groups', 'flags']
+
+    def __init__(self, groups=None, flags=0):
+        self.groups = groups or {}
+        self.flags = flags
+
+
+class CharactersBuilder(object):
+    """Helper object that allows to configure `characters` strategy with
+    various unicode categories and characters. Also allows negation of
+    configured set.
+
+    :param negate: If True, configure :func:`hypothesis.strategies.characters`
+        to match anything other than configured character set
+    :param flags: Regex flags. They affect how and which characters are matched
+
+    """
+
+    def __init__(self, negate=False, flags=0):
+        self._categories = set()
+        self._whitelist_chars = set()
+        self._blacklist_chars = set()
+        self._negate = negate
+        self._ignorecase = flags & re.IGNORECASE
+        self._unicode = not bool(flags & re.ASCII) \
+            if PY3 else bool(flags & re.UNICODE)
+
+    @property
+    def strategy(self):
+        """Returns resulting strategy that generates configured char set."""
+        max_codepoint = None if self._unicode else 127
+
+        strategies = []
+        if self._negate:
+            if self._categories or self._whitelist_chars:  # pragma: no branch
+                strategies.append(
+                    hs.characters(
+                        blacklist_categories=self._categories | set(
+                            ['Cc', 'Cs']),
+                        blacklist_characters=self._whitelist_chars,
+                        max_codepoint=max_codepoint,
+                    )
+                )
+            if self._blacklist_chars:
+                strategies.append(
+                    hs.sampled_from(
+                        sorted(self._blacklist_chars - self._whitelist_chars)
+                    )
+                )
+        else:
+            if self._categories or self._blacklist_chars:
+                strategies.append(
+                    hs.characters(
+                        whitelist_categories=self._categories,
+                        blacklist_characters=self._blacklist_chars,
+                        max_codepoint=max_codepoint,
+                    )
+                )
+            if self._whitelist_chars:
+                strategies.append(
+                    hs.sampled_from(
+                        sorted(self._whitelist_chars - self._blacklist_chars)
+                    )
+                )
+
+        return hs.one_of(*strategies) if strategies else hs.just(u'')
+
+    def add_category(self, category):
+        """Add unicode category to set.
+
+        Unicode categories are strings like 'Ll', 'Lu', 'Nd', etc. See
+        `unicodedata.category()`
```

Comment:
This docstring seems to be unconnected to what the function actually does. The arguments to this are `sre_parse` category objects, which are not unicode categories.